### PR TITLE
Check sync words count to be 24 (direct 1.40.x PR)

### PR DESCRIPTION
**Pull request by AlexeyBarabash**
Resolves https://github.com/brave/brave-browser/issues/23206

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
   1. Take Nightly browser (1.41.x)
   2. Create a sync chain, copy sync code for desktop into clipboard
   3. Take Beta (1.40.x)
   4. Join the sync chain with words from the clipboard
   5. It should give an error message

### CODE REVIEW

**Review by Brandon-T — APPROVED** [review]

File: components/brave_sync/crypto/crypto.cc [hunk 1]
```diff
@@ -158,6 +159,15 @@ bool PassphraseToBytes32(const std::string& passphrase,
...
 
 bool IsPassphraseValid(const std::string& passphrase) {
+  std::vector<std::string> words = base::SplitString(
+      passphrase, " ", base::WhitespaceHandling::TRIM_WHITESPACE,
+      base::SplitResult::SPLIT_WANT_NONEMPTY);
+  size_t num_words = words.size();
+  static constexpr size_t kPureWordsCount = 24u;
+  if (num_words != kPureWordsCount) {
+    return false;
+  }
```
Comment (darkdh) [inline components/brave_sync/crypto/crypto.cc:169]: We should put a comment saying this check is dedicated for old client to reject sync code from new client which has time limited code (25 words)

**Review by darkdh — APPROVED** [review]

File: components/brave_sync/crypto/crypto.cc (hunk 1 above)
Comment (AlexeyBarabash) [reply components/brave_sync/crypto/crypto.cc:171]: thanks, fixed at https://github.com/brave/brave-core/compare/d45af0114142144b9437d2980dc27bb52d734b7c..3bf3aa99a87396d1891508915517ffcb52442e08

**Comment by AlexeyBarabash** [comment]
rebased to pass CI

**Comment by AlexeyBarabash** [comment]
rebased again to pass CI

**Comment by kjozwiak** [comment]
Verification **PASSED** on both `Ubuntu 22.04 x64` & `Win 11 x64` using the following build that was provided by @AlexeyBarabash:

```
Brave | 1.40.84 Chromium: 102.0.5005.78 (Developer Build) (64-bit)
--- | ---
Revision | df6dbb5a9fd82af3f567198af2eb5fb4876ef99c-refs/branch-heads/5005_59@{#3}
```

* installed `1.40.84 Chromium: 102.0.5005.78` which includes the above fix on both `Win 11 x64` & `PopOS 22.04 x64`
* generated a sync chain using `1.41.40 Chromium: 102.0.5005.78` which includes the new time-based sync codes
* ensured that the sync code generated from `1.41.40 Chromium: 102.0.5005.78` couldn't be restored as per the following:
  * ensured that an error message appears when restoring on `Win 11 x64` using `1.40.84 Chromium: 102.0.5005.78`
  * ensured that an error message appears when restoring on `PopOS 22.04 x64` using `1.40.84 Chromium: 102.0.5005.78`

**`Example`** | **`Example`**
---------------|------------------
syncErrorLinux | syncErrorWindows

Also ensured that sync codes that are not time-based can still be restored into  `1.40.84 Chromium: 102.0.5005.78`  without any issues. STR/Cases used:

* generated a sync chain using `1.39.111 Chromium: 102.0.5005.61` which doesn't have time-based sync backup codes
* ensured that the sync code generated from  `1.39.111 Chromium: 102.0.5005.61` can be restored as per the following:
  * ensured that the `PopOS 22.04 x64` device was added into the sync chain without issues
  * ensured that the `Win 11 x64` device was added into the sync chain without issues

image

**Review by kjozwiak — APPROVED** [review]
Uplift into `1.40.x` approved 👍

**Pull request merged**